File: Medium/1792. Maximum Average Pass Ratio.py
from heapq import heappush, heapreplace
from typing import List


class Solution:
    def maxAverageRatio(self, classes: List[List[int]], extra_students: int) -> float:
        N = len(classes)
        sum = 0
        heap = []
        for p, q in classes:
            sum += p / q
            heappush(heap, ((p - q) / (q * (q + 1)), p, q))
        for _ in range(extra_students):
            r, p, q = heap[0]
            if r == 0:
                break
            sum -= r
            nr = (p - q) / ((q + 1.0) * (q + 2.0))
            heapreplace(heap, (nr, p + 1, q + 1))
        return sum / N

File: Medium/1792. Maximum Average Pass Ratio.java
import java.util.PriorityQueue;

/**
 * Solution to maximize the average pass ratio after assigning extra students.
 */
class Solution {
    /**
     * Calculates the maximum average pass ratio after distributing extra students.
     *
     * @param classes       2D array where classes[i][0] is passed students,
     *                      classes[i][1] is total students
     * @param extraStudents number of extra students to assign
     * @return maximum average pass ratio
     */
    public double maxAverageRatio(int[][] classes, int extraStudents) {
        int n = classes.length;
        // Max heap based on the increase in pass ratio if an extra student is added
        PriorityQueue<double[]> maxHeap = new PriorityQueue<>(
                (a, b) -> Double.compare(b[2], a[2]));
        double totalRatio = 0.0;

        // Initialize the heap with each class's current state and potential increase
        for (int[] c : classes) {
            double pass = c[0];
            double total = c[1];
            double increment = (pass + 1) / (total + 1) - pass / total;
            totalRatio += pass / total;
            maxHeap.offer(new double[] { pass, total, increment });
        }

        // Assign each extra student to the class with the highest potential increase
        for (int i = 0; i < extraStudents; i++) {
            double[] curr = maxHeap.poll();
            double pass = curr[0] + 1;
            double total = curr[1] + 1;
            double increment = (pass + 1) / (total + 1) - pass / total;
            totalRatio += curr[2];
            maxHeap.offer(new double[] { pass, total, increment });
        }

        return totalRatio / n;
    }
}